File: solutions/scala/saddle-points/1/src/main/scala/SaddlePoints.scala
case class Matrix(matrix: List[List[Int]]) {

  def saddlePoints: Set[(Int, Int)] = {
    (for (row <- matrix;
          value <- row;
          index = row.indexOf(value)
          if row.max == matrix.map(_(index)).min)
      yield (matrix.indexOf(row), index))
      .toSet
  }
}

File: solutions/typescript/saddle-points/1/saddle-points.ts
export function saddlePoints(matrix: number[][]): { row: number; column: number }[] {
  const points = []

  for (let row = 0; row < matrix.length; row++) {
    const rowMaximum = Math.max(...matrix[row])
    for (let column = 0; column < matrix[row].length; column++) {
      const columnMinimium = Math.min(...matrix.flatMap(row => row[column]))

      if (matrix[row][column] === columnMinimium && matrix[row][column] === rowMaximum)
        points.push({ column, row })
    }  
  }
  
  return points.map(({ column, row }) => ({ column: column + 1, row: row + 1 }))
}